### PR TITLE
Improve handling of QTBUG-57971 (ListModel initializing roles on first insertion)

### PR DESCRIPTION
**Pull request by micieslak**
# Description #

So far the provided workaround was missing some corner cases when there is more than one SFPM (or other proxies) in a chain, providing some custom roles like proxy roles in SFPM. Detailed explanation provided as a comment directly in the source.

Additional information about failing corner cases can be found here: https://github.com/status-im/status-desktop/issues/16310

### CODE REVIEW

File: qqmlsortfilterproxymodel.cpp [hunk 1]
```diff
@@ -380,10 +380,39 @@ void QQmlSortFilterProxyModel::resetInternalData()
...
+    // (at least proxy roles). It means that on first insertion right after
+    // SFPM creation or after source model reset it's necessary to re-initialize
+    // role names if the source was empty before insertion.
+
+    if (sourceModel && sourceModel->rowCount() == 0)
+        connect(sourceModel, &QAbstractItemModel::rowsInserted, this,
+                &QQmlSortFilterProxyModel::initRoles, Qt::UniqueConnection);
+
+    if (sourceModel) {
+        connect(sourceModel, &QAbstractItemModel::modelReset, this, [sourceModel, this]() {
```
Comment (alexjba) [inline qqmlsortfilterproxymodel.cpp:413]: We should also disconnect this after the first insertion, right?

Comment (micieslak) [reply qqmlsortfilterproxymodel.cpp:413]: No, this connection must stay here forever, otherwise cases like https://github.com/status-im/status-desktop/issues/16310#issuecomment-2348491548 won't work. The reset may be emitted because the source model has own source changed. And source of source may be e.g. `ListModel` with not initialized roles. So we must assume that after every source reset the workaround procedure must be connected again if source is empty.

**Review by caybro — COMMENTED** [review]
Looks good

Need a regression test; since we know the scenarios under which this happens

**Comment by micieslak** [comment]
> Looks good
> 
> Need a regression test; since we know the scenarios under which this happens

Once it's merged we can test the app in a https://github.com/status-im/status-desktop/pull/16321 where the `SFPM` is bumped.

**Comment by caybro** [comment]
> > Looks good
> > 
> > Need a regression test; since we know the scenarios under which this happens
> 
> Once it's merged we can test the app in a https://github.com/status-im/status-desktop/pull/16321 where the `SFPM` is bumped.

Yeah sure, realized too late this is a different repo 😊

File: qqmlsortfilterproxymodel.cpp (hunk 1 above)
Comment (alexjba) [reply qqmlsortfilterproxymodel.cpp:413]: Got it, thanks! In this case I think we should disconnect on sourceModel change

Comment (micieslak) [reply qqmlsortfilterproxymodel.cpp:413]: Ahh, you are right! Your statement is true even for the original implementation. The connection to `rowsInserted` should be also disconnected on source change.

Comment (micieslak) [reply qqmlsortfilterproxymodel.cpp:413]: Thank you @alexjba for your discovery, there is excerpt exploiting that problem:

```qml
import QtQuick 2.15
import QtQuick.Controls 2.15

import StatusQ.Core.Utils 0.1

import SortFilterProxyModel 0.2

Item {
    id: root

    ListModel {
        id: emptyModel
    }

    ListModel {
        id: emptyModel2
    }

    SortFilterProxyModel {
        id: sfpm

        sourceModel: emptyModel

        proxyRoles: ExpressionRole {
            name: "customRole"
            expression: "X" + model.index
        }
    }

    ListView {
        id: lv

        anchors.fill: parent

        model: sfpm

        delegate: Text {
            text: model.text + " | " + model.customRole
        }
    }

    Button {
        anchors.centerIn: parent

        text: "ADD"

        onClicked: {
            sfpm.sourceModel = emptyModel2

            emptyModel.append({ text: "d" })
            emptyModel2.append({ text: "d" })
        }
    }
}
```

Inserting into model that was a source (but is not currently) triggers the `initRoles` and disconnects the slot from currently used model :scream:. So when inserting into current source, roles are not initialized properly... :man_facepalming:

Comment (micieslak) [reply qqmlsortfilterproxymodel.cpp:413]: Another example, this one exploiting lack of disconnection from `modelReset`:

```qml
import QtQuick 2.15
import QtQuick.Controls 2.15

import StatusQ.Core.Utils 0.1

import SortFilterProxyModel 0.2

Item {
    id: root

    ListModel {
        id: emptyModel
    }

    ListModel {
        id: emptyModel2
    }

    ListModel {
        id: emptyModel3
    }

    SortFilterProxyModel {
        id: sfpm1

        sourceModel: emptyModel

        proxyRoles: ExpressionRole {
            name: "customRole"
            expression: "X" + model.index
        }
    }

    SortFilterProxyModel {
        id: sfpm3

        sourceModel: emptyModel3

        proxyRoles: ExpressionRole {
            name: "customRole"
            expression: "X-" + model.index
        }
    }


    SortFilterProxyModel {
        id: sfpm2

        sourceModel: sfpm1
    }

    ListView {
        id: lv

        anchors.fill: parent

        model: sfpm2

        delegate: Text {
            text: model.text + " | " + model.customRole
        }
    }

    Button {
        anchors.centerIn: parent

        text: "ADD"

        onClicked: {
            sfpm2.sourceModel = sfpm3

            sfpm1.sourceModel = emptyModel2

            emptyModel2.append({ text: "d" })
            emptyModel3.append({ text: "d" })
        }
    }
}
```

Comment (micieslak) [reply qqmlsortfilterproxymodel.cpp:413]: I provided a fix. @alexjba @caybro please take one more look.

**Review by caybro — APPROVED** [review]
LGTM!

**Review by alexjba — APPROVED** [review]
Awesome! Thank you 🍻

**Comment by micieslak** [comment]
Ideally we should add proper unit tests for those corner cases. However now it's not easy as all helpers we are using for proxy models testing are located in `status-desktop` repository and are not available for `SFPM`. For now, I'm placing here list of scenarios failing before the fix was introduces:

https://github.com/status-im/status-desktop/issues/16310#issuecomment-2348491548
https://github.com/status-im/status-desktop/issues/16310#issuecomment-2348523535
https://github.com/status-im/SortFilterProxyModel/pull/5#discussion_r1758761147
https://github.com/status-im/SortFilterProxyModel/pull/5#discussion_r1758977438

**Comment by caybro** [comment]
> Ideally we should add proper unit tests for those corner cases. However now it's not easy as all helpers we are using for proxy models testing are located in `status-desktop` repository and are not available for `SFPM`. For now, I'm placing here list of scenarios failing before the fix was introduces:
> 
> [status-im/status-desktop#16310 (comment)](https://github.com/status-im/status-desktop/issues/16310#issuecomment-2348491548) [status-im/status-desktop#16310 (comment)](https://github.com/status-im/status-desktop/issues/16310#issuecomment-2348523535) [#5 (comment)](https://github.com/status-im/SortFilterProxyModel/pull/5#discussion_r1758761147) [#5 (comment)](https://github.com/status-im/SortFilterProxyModel/pull/5#discussion_r1758977438)

There already some tests under `vendor/SortFilterProxyModel/tests`; we could add new ones there and integrate the whole suite into `ui-tests`

**Comment by micieslak** [comment]
> > Ideally we should add proper unit tests for those corner cases. However now it's not easy as all helpers we are using for proxy models testing are located in `status-desktop` repository and are not available for `SFPM`. For now, I'm placing here list of scenarios failing before the fix was introduces:
> > [status-im/status-desktop#16310 (comment)](https://github.com/status-im/status-desktop/issues/16310#issuecomment-2348491548) [status-im/status-desktop#16310 (comment)](https://github.com/status-im/status-desktop/issues/16310#issuecomment-2348523535) [#5 (comment)](https://github.com/status-im/SortFilterProxyModel/pull/5#discussion_r1758761147) [#5 (comment)](https://github.com/status-im/SortFilterProxyModel/pull/5#discussion_r1758977438)
> 
> There already some tests under `vendor/SortFilterProxyModel/tests`; we could add new ones there and integrate the whole suite into `ui-tests`

Good point, I created a ticket: https://github.com/status-im/status-desktop/issues/16343
